src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1882,6 +1906,204 @@ MutableBigInteger sqrt() {
+
+        // Normalize
+        MutableBigInteger x = this;
+        int shift = Integer.numberOfLeadingZeros(x.value[x.offset]) & ~1; // shift must be even
+        if (shift != 0) {
+            x = new MutableBigInteger(x);
+            x.primitiveLeftShift(shift);
+        }
+
+        MutableBigInteger[] sqrtRem = x.sqrtRemZimmermann(x.intLen, needRemainder);
+
+        // Unnormalize
+        if (shift != 0) {
+            final int halfShift = shift >> 1;
+            if (needRemainder && !sqrtRem[1].isZero()) {
+                final int s0 = sqrtRem[0].value[sqrtRem[0].offset + sqrtRem[0].intLen - 1]
+                        & (-1 >>> -halfShift); // Remove excess bits
+                if (s0 != 0) { // An optimization
+                    // s0 is at most 15 bit long
+                    MutableBigInteger doubleProd = new MutableBigInteger();
+                    sqrtRem[0].mul(s0 << 1, doubleProd);
+
+                    sqrtRem[1].add(doubleProd);
+                    sqrtRem[1].subtract(new MutableBigInteger(s0 * s0));
+                }
+                sqrtRem[1].primitiveRightShift(shift);
+            }
+            sqrtRem[0].primitiveRightShift(halfShift);
+        }
+        return sqrtRem;
+    }
+
+    /**
+     * Assume {@code 2 <= len <= intLen && Integer.numberOfLeadingZeros(value[offset]) <= 1}
+     * @implNote The implementation is based on Zimmermann's works available
+     * <a href="https://inria.hal.science/inria-00072854/en/">  here</a> and
+     * <a href="https://www.researchgate.net/publication/220532560_A_proof_of_GMP_square_root">  here</a>
+     */
+    private MutableBigInteger[] sqrtRemZimmermann(int len, boolean needRemainder) {
+        if (len == 2) { // Base case
+            long x = ((value[offset] & LONG_MASK) << 32) | (value[offset + 1] & LONG_MASK);
+            long s = (long) Math.sqrt(x > 0 ? x : x + 0x1p64);
+            if (s > LONG_MASK || Long.compareUnsigned(x, s * s) < 0)
+                s--;
+
+            // Allocate sufficient space to hold the normalized final square root
+            MutableBigInteger sqrt = new MutableBigInteger(new int[(intLen + ((-intLen) & 3) + 1) >> 1]);
+
+            // Place the partial square root
+            sqrt.intLen = 1;
+            sqrt.value[0] = (int) s;
+
+            // The first invocation is never a base case, so the remainder is needed
+            return new MutableBigInteger[] { sqrt, new MutableBigInteger(x - s * s) };
+        }
+
+        // Recursive step (len >= 3)
+
+        // Normalize
+        /* For speed, the normalization is performed only "logically",
+         * which means that the contents of the input are not changed,
+         * but only its logical length, which is stored in the parameter len.
+         * The length is normalized to a mutiple of 4 because, in this way,
+         * the input can be always split in blocks of words without twiddling with bits.
+         */
```

Comment:
What's substantially different here from the Bertot, Magaud, Zimmermann paper?
This is not explained in detail, so it is way harder to understand than the paper.

Reply:
In the paper, the number is normalized to an even number of words... but this is not done in the pseudocode, since the pseudocode assumes that the number is already normalized, so actually there's no substantial difference...

Reply:
Yes, normalization is done in §3.2. Will have another look.